src/pip/_internal/commands/install.py

```diff
@@ -605,84 +601,139 @@ def _warn_about_conflicts(self, conflict_details, new_resolver):
 
 
 def get_lib_location_guesses(
-        user=False,  # type: bool
-        home=None,  # type: Optional[str]
-        root=None,  # type: Optional[str]
-        isolated=False,  # type: bool
-        prefix=None  # type: Optional[str]
+    user=False,  # type: bool
+    home=None,  # type: Optional[str]
+    root=None,  # type: Optional[str]
+    isolated=False,  # type: bool
+    prefix=None,  # type: Optional[str]
 ):
-    # type:(...) -> List[str]
+    # type: (...) -> List[str]
     scheme = distutils_scheme('', user=user, home=home, root=root,
                               isolated=isolated, prefix=prefix)
     return [scheme['purelib'], scheme['platlib']]
 
 
-def site_packages_writable(root, isolated):
-    # type: (Optional[str], bool) -> bool
-    return all(
-        test_writable_dir(d) for d in set(
-            get_lib_location_guesses(root=root, isolated=isolated))
-    )
+def site_packages_writable(
+    user=False,  # type: bool
+    root=None,  # type: Optional[str]
+    isolated=False,  # type: bool
+    prefix=None,  # type: Optional[str]
+):
+    # type: (...) -> bool
+    return all(map(test_writable_dir, set(get_lib_location_guesses(
+        user=user, root=root, isolated=isolated, prefix=prefix,
+    ))))
 
 
 def decide_user_install(
-    use_user_site,  # type: Optional[bool]
+    use_user_site=None,  # type: Optional[bool]
     prefix_path=None,  # type: Optional[str]
     target_dir=None,  # type: Optional[str]
     root_path=None,  # type: Optional[str]
     isolated_mode=False,  # type: bool
 ):
     # type: (...) -> bool
-    """Determine whether to do a user install based on the input options.
+    """Determine whether to do a user install.
```

Comment:
This appears to be a complete rewrite of `decide_user_install` - I can't follow how the logic changes from the old to the new version, the differences are too major to find any obvious points of similarity. It would be much easier to review this PR if it was possible to understand what actually changed here.

Also, given that the PR is "fail early when install path is not writeable", I don't understand why you're changing this function in any case. The changes here seem unrelated to the purpose of the PR, and if they *are* necessary, they should be rewritten to make the reasoning clearer.

I'm not going to review this function in any further detail, as I don't believe I can say anything meaningful.

Reply:
> the differences are too major to find any obvious points of similarity. It would be much easier to review this PR if it was possible to understand what actually changed here.

I strongly agree. Part of the reason that I've not revisited this PR personally is that working through those changes (and the effects of them) is pretty complicated, and not something I'm comfortable doing in even a 20 minute sitting, due to how nuanced the logic around this entire area is and how drastic the changes in this PR are.

Reply:
+1 I totally relate to this and I normally request that submitters distill as many small and atomic changes into separate PRs as possible...

Reply:
@McSinyx if you were to learn just one important thing from this review it must be this :point_up:

Reply:
Thank you all for the suggestions and sorry for requesting reviews and running away :see_no_evil:  Given I'm more experienced with submitting code for peer review now, I'll try to redo this one for it to make some sense, likely within the next few days.